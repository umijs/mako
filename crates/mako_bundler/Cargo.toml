[package]
name = "mako_bundler"
version = "0.0.0"
edition = "2021"
repository = "https://github.com/umijs/mako"
license = "MIT"

[dependencies]
maplit = "1.0.2"
tokio = { workspace = true, features = ["full"] }
swc_ecma_ast = { version = "0.102.5", features = ["rkyv-impl"] }
swc_ecma_parser = { version = "0.132.6" }
swc_common = { version = "0.30.5", features = ["concurrent"] }
swc_ecma_visit = { version = "0.88.5" }
swc_ecma_codegen = { version = "0.137.6" }
swc_ecma_transforms = { version = "0.216.4", features = [
  "module",
  "typescript",
  "react",
  "concurrent",
] }
swc_ecma_utils = { version = "0.115.8" }
swc_atoms = { version = "0.4.43" }
relative-path = "1.8.0"
regex = "1"
lazy_static = "1.4.0"
base64 = "0.21.0"
petgraph = "0.6.3"
md5 = "0.7.0"
serde = { version = "1.0.160", features = ["derive"] }
serde_json = "1.0.96"
config = "0.13.3"
nodejs-resolver = "0.0.78"
lightningcss = "1.0.0-alpha.41"
swc_css_ast = "0.136.6"
swc_css_codegen = "0.146.6"
swc_css_parser = "0.145.6"
swc_css_visit = "0.135.6"
rayon = "1.7.0"

[dev-dependencies]
insta = { version = "1.29.0", features = ["yaml"] }
