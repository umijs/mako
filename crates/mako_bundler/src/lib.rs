#![feature(box_patterns)]

use compiler::Compiler;

use crate::plugins::node_polyfill::get_node_builtins;

pub mod build;
pub mod chunk;
pub mod chunk_graph;
pub mod compiler;
pub mod config;
pub mod context;
pub mod generate;
pub mod module;
pub mod module_graph;
pub mod plugin;
pub mod plugins;
pub mod utils;

pub fn run() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() <= 1 {
        panic!("Please specify the root directory of the project");
    }

    // config
    let root = std::env::current_dir()
        .unwrap()
        .join(&args[1])
        .to_string_lossy()
        .to_string();
    let mut config = config::Config::from_literal_str(
        format!(
            r#"
{{
    "externals": {{}},
    "root": "{}",
    "entry": {{ "index": "index.tsx" }}
}}
    "#,
            root
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();

    // TODO: move this to plugin
    // node polyfills
    let builtins = get_node_builtins();
    for name in builtins.iter() {
        config.externals.insert(name.to_string(), name.to_string());
    }

    // compiler_origin::run_compiler(config);
    let mut compiler = Compiler::new(config);
    compiler.run();

    println!("âœ… DONE");
}
