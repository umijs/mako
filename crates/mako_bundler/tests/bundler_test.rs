use mako_bundler::{
    build::build::BuildParam, compiler::Compiler, config::Config, generate::generate::GenerateParam,
};

#[test]
fn normal() {
    let (output, ..) = test_files("normal".into());
    assert_debug_snapshot!(output);
}

#[test]
fn multiple_files() {
    let (output, mut compiler, ..) = test_files("multiple".into());
    assert_debug_snapshot!(output);
    let (orders, _) = compiler.context.module_graph.topo_sort();
    assert_debug_snapshot!(&orders);
    assert_display_snapshot!(&compiler.context.module_graph);
}

#[test]
fn replace_env() {
    let (output, ..) = test_files("env".into());
    assert_debug_snapshot!(output);
}

#[test]
fn chunk() {
    let (output, compiler, ..) = test_files("chunks".into());
    assert_debug_snapshot!(output);
    assert_display_snapshot!(compiler.context.chunk_graph);
}

#[allow(clippy::useless_format)]
fn test_files(name: String) -> (Vec<String>, Compiler, String) {
    let cwd = std::env::current_dir()
        .unwrap()
        .join("tests/fixtures")
        .join(&name)
        .to_string_lossy()
        .to_string();
    let mut config = Config::from_literal_str(
        format!(
            r#"
{{
    "entry": {{
        "entry": "entry.js"
    }},
    "root": "{}"
}}
            "#,
            cwd,
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();
    let mut compiler = Compiler::new(config);
    compiler.build(&BuildParam { files: None });
    let generate_result = compiler.generate(&GenerateParam { write: false });
    let output = generate_result.output_files[0].__output.clone();
    (output, compiler, cwd)
}

#[macro_export]
macro_rules! assert_display_snapshot {
    ($value:expr) => {{
        let cwd = std::env::current_dir()
            .unwrap()
            .to_string_lossy()
            .to_string();
        let value = format!("{}", $value).replace(&cwd, "<CWD>");
        insta::assert_snapshot!(insta::_macro_support::AutoName, value, stringify!($value));
    }};
}

#[macro_export]
macro_rules! assert_debug_snapshot {
    ($value:expr) => {{
        let cwd = std::env::current_dir()
            .unwrap()
            .to_string_lossy()
            .to_string();
        let value = format!("{:#?}", $value).replace(&cwd, "<CWD>");
        insta::assert_snapshot!(insta::_macro_support::AutoName, value, stringify!($value));
    }};
}
