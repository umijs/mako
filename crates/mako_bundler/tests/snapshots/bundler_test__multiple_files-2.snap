---
source: crates/mako_bundler/tests/bundler_test.rs
expression: "&orders"
---
[
    ModuleId {
        id: "<CWD>/tests/fixtures/multiple/entry.js",
    },
    ModuleId {
        id: "<CWD>/tests/fixtures/multiple/one.js",
    },
    ModuleId {
        id: "<CWD>/tests/fixtures/multiple/two.js",
    },
    ModuleId {
        id: "<CWD>/tests/fixtures/multiple/three.js",
    },
]
